sysfs/class_infiniband.go

```diff
@@ -0,0 +1,372 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !windows
+
+package sysfs
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/procfs/internal/util"
+)
+
+const infinibandClassPath = "class/infiniband"
+
+// InfiniBandCounters contains counter values from files in
+// /sys/class/infiniband/<Name>/ports/<Port>/counters or
+// /sys/class/infiniband/<Name>/ports/<Port>/counters_ext
+// for a single port of one InfiniBand device.
+type InfiniBandCounters struct {
+	LegacyPortMulticastRcvPackets  *uint64 // counters_ext/port_multicast_rcv_packets
```

Comment:
Do we have to have the legacy versions as well? Would it be possible to look for both new/legacy and only use legacy if the new version is absent?

Reply:
We run a mixture of kernel versions where I work, and I've had to do some elaborate metric relabelling to make our IB metric names conform to a common standard, so that they're supported by a one-size-fits-all Grafana dashboard. Brace yourself for a verbose reason why...

Kernels 4.5 and later merge the `counters_ext` IB counters in sysfs into the `counters` directory. This effectively replaces the 32-bit counters that were previously exposed in the `counters` directory with the 64-bit counters from the `counters_ext` directory.

Unfortunately, InfiniBand counters do not automatically roll over, so when the 32-bit counters reach 4 billion, they simply latch at that value (which is obviously bad for `rate` / `irate` queries).

Node_exporter doesn't really handle this in a sane way, so on pre-4.5 kernels it exposes 32-bit counters as e.g. `node_infiniband_port_packets_transmitted_total` and 64-bit counters as
e.g. `node_infiniband_legacy_packets_transmitted_total`. This makes it difficult to have unified dashboards / alerting rules which can handle both pre- and post-4.5 kernels.

To cut a long story short, I ended up configuring the following relabelling rules, which rename the legacy metrics if exposed, and overwrite the "non-legacy" metrics (which will be 32-bit, and thus useless, if legacy metrics are present):
```yaml
  # Rename node_infiniband_legacy_* (pre-MLNX_OFED 4.0) to unify PromQL queries
  - source_labels: [__name__]
    target_label: __name__
    regex: node_infiniband_legacy_((?:data|packets)_.+)
    replacement: node_infiniband_port_$1
  - source_labels: [__name__]
    target_label: __name__
    regex: node_infiniband_legacy_((?:multicast|unicast)_.+)
    replacement: node_infiniband_$1
```
In short - it would be nice if node_exporter simply exposed the widest counters available, regardless of kernel / OFED version. However, that is something that needs to happen in node_exporter, not procfs, and it may be desirable to still expose both legacy (32-bit) and modern (64-bit) counters, at least in procfs.

Reply:
Thanks for the answer @dswarbrick. Therefore I mark this conversation as resolved without changing the code.